File: C++/Atletas.cpp
#include <iostream>
#include <string.h>
using namespace std;
#define MaxAtletas 30
#define MaxJueces 30

// Estructura para la información de los jueces
struct Juez
{
  char nombre[50];
  char pais[50];
  float puntaje;
};

// Estructura para la información de los saltos de un atleta
struct Salto
{
  char dificultad; // A, B, C o D
  float puntaje;
};

// Estructura para la información de un atleta
struct Atleta
{
  char pais[50];
  char nombre[50];
  int edad;
  Salto intentos[3];
  float puntajeTotal;
};

// Función para cargar la información de un atleta
void CargarAtleta(Atleta atleta)
{
  cout << "Ingrese el país del atleta: ";
  cin.getline(atleta.pais, 50);

  cout << "Ingrese el nombre del atleta: ";
  cin.getline(atleta.nombre, 50);

  cout << "Ingrese la edad del atleta: ";
  cin >> atleta.edad;
  cin.ignore(); // Limpiar el buffer de entrada

  for (int i = 0; i < 3; i++)
  {
    cout << "Salto #" << i + 1 << ":" << endl;
    cout << "Dificultad del salto (A, B, C o D): ";
    cin >> atleta.intentos[i].dificultad;
    cout << "Puntaje otorgado (0 a 10): ";
    cin >> atleta.intentos[i].puntaje;
    cin.ignore(); // Limpiar el buffer de entrada
  }
}

// Función para calcular el puntaje total de un atleta
void CalcularPuntaje(Atleta atleta)
{
  float puntajes[3];
  for (int i = 0; i < 3; i++)
  {
    puntajes[i] = atleta.intentos[i].puntaje;
  }

  // Eliminar el puntaje más alto y el puntaje más bajo
  float minPuntaje = puntajes[0];
  float maxPuntaje = puntajes[0];
  int minIndex = 0;
  int maxIndex = 0;

  for (int i = 1; i < 3; i++)
  {
    if (puntajes[i] < minPuntaje)
    {
      minPuntaje = puntajes[i];
      minIndex = i;
    }

    if (puntajes[i] > maxPuntaje)
    {
      maxPuntaje = puntajes[i];
      maxIndex = i;
    }
  }

  puntajes[minIndex] = 0; // Eliminar el puntaje más bajo
  puntajes[maxIndex] = 0; // Eliminar el puntaje más alto

  // Calcular el puntaje total
  float puntajeTotal = 0;
  for (int i = 0; i < 3; i++)
  {
    puntajeTotal += puntajes[i];
  }

  // Multiplicar por el grado de dificultad
  switch (atleta.intentos[0].dificultad)
  {
  case 'A':
    puntajeTotal *= 0.45;
    break;
  case 'B':
    puntajeTotal *= 0.70;
    break;
  case 'C':
    puntajeTotal *= 0.85;
    break;
  case 'D':
    puntajeTotal *= 1.2;
    break;
  default:
    cout << "Dificultad no válida." << endl;
    return;
  }

  atleta.puntajeTotal = puntajeTotal;
}

// Función para mostrar atletas que no realizaron saltos con dificultad B
void MostrarAtletasSinSaltoBDificultad(Atleta atletas[], int numAtletas)
{
  cout << "Atletas que no realizaron saltos con dificultad B:" << endl;
  for (int i = 0; i < numAtletas; i++)
  {
    bool tieneSaltoBDificultad = false;
    for (int j = 0; j < 3; j++)
    {
      if (atletas[i].intentos[j].dificultad == 'B')
      {
        tieneSaltoBDificultad = true;
        break;
      }
    }

    if (!tieneSaltoBDificultad)
    {
      cout << "Nombre: " << atletas[i].nombre << ", País: " << atletas[i].pais << endl;
    }
  }
}

// Función para mostrar jueces que otorgaron puntaje mayor de 8.75 a su mismo país en el primer salto
void MostrarJuecesPuntajeMayor(Atleta atletas[], Juez jueces[], int numAtletas, int numJueces)
{
  cout << "Jueces que otorgaron puntaje mayor de 8.75 a su mismo país en el primer salto:" << endl;
  for (int i = 0; i < numAtletas; i++)
  {
    for (int j = 0; j < numJueces; j++)
    {
      // Busamos el pais y la comparacion del puntaje
      if (strcmp(atletas[i].pais, jueces[j].pais) == 0 && atletas[i].intentos[0].puntaje > 8.75)
      {
        cout << "Nombre del Juez: " << jueces[j].nombre << ", País del Juez: " << jueces[j].pais << endl;
      }
    }
  }
}

int main()
{
  int cantidadAtletas, cantidadJueces;
  Atleta atletas[MaxAtletas];
  Juez jueces[MaxJueces];

  printf("Ingrese cantidad de atletas (maximo 30): ");
  cin >> cantidadAtletas;
  do
  {
    if ((cantidadAtletas < 0) || (cantidadAtletas > 30))
    {
      printf("Ingrese una cantidad valida: ");
      cin >> cantidadAtletas;
    }
  } while ((cantidadAtletas < 0) || (cantidadAtletas > 30));

  printf("Ingrese cantidad de jueces (maximo 9) : ");
  cin >> cantidadJueces;
  do
  {
    if ((cantidadJueces < 0) || (cantidadJueces > 9))
    {
      printf("Ingrese una cantidad valida: ");
      cin >> cantidadJueces;
    }
  } while ((cantidadJueces < 0) || (cantidadJueces > 9));

  // Cargar información de los atletas
  for (int i = 0; i < cantidadAtletas; i++)
  {
    cout << "Datos del atleta #" << i + 1 << ":" << endl;
    CargarAtleta(atletas[i]);
    CalcularPuntaje(atletas[i]);
    cout << "------------------------------------" << endl;
  }

  // Cargar información de los jueces
  for (int i = 0; i < cantidadJueces; i++)
  {
    cout << "Datos del juez #" << i + 1 << ":" << endl;
    cout << "Ingrese el nombre del juez: ";
    cin.getline(jueces[i].nombre, 50);

    cout << "Ingrese el país del juez: ";
    cin.getline(jueces[i].pais, 50);

    cout << "Ingrese el puntaje otorgado por el juez (0 a 10): ";
    cin >> jueces[i].puntaje;
    cin.ignore(); // Limpiar el buffer de entrada
    cout << "------------------------------------" << endl;
  }

  // Ejemplo de uso de funciones
  MostrarAtletasSinSaltoBDificultad(atletas, cantidadAtletas);
  cout << "------------------------------------" << endl;
  MostrarJuecesPuntajeMayor(atletas, jueces, cantidadAtletas, cantidadJueces);

  return 0;
}

File: Python/Atletas.py


# Estructura para la información de los jueces
class Juez:
    def __init__(self):
        self.nombre = ''
        self.pais = ''
        self.puntaje = 0.0

# Estructura para la información de los saltos de un atleta
class Salto:
    def __init__(self):
        self.dificultad = ''  # A, B, C o D
        self.puntaje = 0.0

# Estructura para la información de un atleta
class Atleta:
    def __init__(self):
        self.pais = ''
        self.nombre = ''
        self.edad = 0
        self.intentos = [Salto() for _ in range(3)]
        self.puntajeTotal = 0.0

# Función para cargar la información de un atleta
def cargar_atleta(atleta):
    atleta.pais = input("Ingrese el país del atleta: ")
    atleta.nombre = input("Ingrese el nombre del atleta: ")
    atleta.edad = int(input("Ingrese la edad del atleta: "))

    for i in range(3):
        print(f"Salto #{i + 1}:")
        atleta.intentos[i].dificultad = input("Dificultad del salto (A, B, C o D): ")
        atleta.intentos[i].puntaje = float(input("Puntaje otorgado (0 a 10): "))

# Función para calcular el puntaje total de un atleta
def calcular_puntaje(atleta):
    puntajes = [salto.puntaje for salto in atleta.intentos]

    # Eliminar el puntaje más alto y el puntaje más bajo
    min_puntaje = min(puntajes)
    max_puntaje = max(puntajes)

    min_index = puntajes.index(min_puntaje)
    max_index = puntajes.index(max_puntaje)

    puntajes[min_index] = 0  # Eliminar el puntaje más bajo
    puntajes[max_index] = 0  # Eliminar el puntaje más alto

    # Calcular el puntaje total
    puntaje_total = sum(puntajes)

    # Multiplicar por el grado de dificultad
    dificultad = atleta.intentos[0].dificultad
    if dificultad == 'A':
        puntaje_total *= 0.45
    elif dificultad == 'B':
        puntaje_total *= 0.70
    elif dificultad == 'C':
        puntaje_total *= 0.85
    elif dificultad == 'D':
        puntaje_total *= 1.2
    else:
        print("Dificultad no válida.")
        return

    atleta.puntajeTotal = puntaje_total

# Función para mostrar atletas que no realizaron saltos con dificultad B
def mostrar_atletas_sin_salto_b_dificultad(atletas, num_atletas):
    print("Atletas que no realizaron saltos con dificultad B:")
    for i in range(num_atletas):
        tiene_salto_b_dificultad = not any(salto.dificultad == 'B' for salto in atletas[i].intentos)

        if not tiene_salto_b_dificultad:
            print(f"Nombre: {atletas[i].nombre}, País: {atletas[i].pais}")

# Función para mostrar jueces que otorgaron puntaje mayor de 8.75 a su mismo país en el primer salto
def mostrar_jueces_puntaje_mayor(atletas, jueces, num_atletas, num_jueces):
    print("Jueces que otorgaron puntaje mayor de 8.75 a su mismo país en el primer salto:")
    for i in range(num_atletas):
        for j in range(num_jueces):
            if atletas[i].pais == jueces[j].pais and atletas[i].intentos[0].puntaje > 8.75:
                print(f"Nombre del Juez: {jueces[j].nombre}, País del Juez: {jueces[j].pais}")

# Función principal
def main():
    cantidad_atletas = int(input("Ingrese cantidad de atletas (máximo 30): "))
    cantidad_jueces = int(input("Ingrese cantidad de jueces (máximo 9): "))

    if not 0 <= cantidad_atletas <= 30 or not 0 <= cantidad_jueces <= 9:
        print("Ingrese una cantidad válida.")
        return

    atletas = [Atleta() for _ in range(cantidad_atletas)]
    jueces = [Juez() for _ in range(cantidad_jueces)]

    # Cargar información de los atletas
    for i in range(cantidad_atletas):
        print(f"Datos del atleta #{i + 1}:")
        cargar_atleta(atletas[i])
        calcular_puntaje(atletas[i])
        print("------------------------------------")

    # Cargar información de los jueces
    for i in range(cantidad_jueces):
        print(f"Datos del juez #{i + 1}:")
        jueces[i].nombre = input("Ingrese el nombre del juez: ")
        jueces[i].pais = input("Ingrese el país del juez: ")
        jueces[i].puntaje = float(input("Ingrese el puntaje otorgado por el juez (0 a 10): "))
        print("------------------------------------")

    # Ejemplo de uso de funciones
    mostrar_atletas_sin_salto_b_dificultad(atletas, cantidad_atletas)
    print("------------------------------------")
    mostrar_jueces_puntaje_mayor(atletas, jueces, cantidad_atletas, cantidad_jueces)

if __name__ == "__main__":
    main()
